tls/s2n_cipher_suites.c

```diff
@@ -1168,13 +1216,175 @@ static int s2n_wire_ciphers_contain(const uint8_t *match, const uint8_t *wire, u
     return 0;
 }
 
+/**
+ * @brief While a potential cipher suite match has been identified, we still need to check if the cipher suite
+ *        can actually be used for the remainder of the connection. These checks ensure that there exists an
+ *        implementation, whether the versions and PSKs are compatible.
+ * 
+ * @param conn s2n_connection context
+ * @param potential_match Potential cipher suite match
+ * @return true Cipher suite match is usable
+ * @return false Cipher suite match is not usable for this connection
+ */
+static bool s2n_cipher_suite_match_is_valid(struct s2n_connection* conn, struct s2n_cipher_suite* potential_match) {
+        /* Never use TLS1.3 ciphers on a pre-TLS1.3 connection, and vice versa */
+        if ((conn->actual_protocol_version >= S2N_TLS13) != (potential_match->minimum_required_tls_version >= S2N_TLS13)) {
+            return false;
+        }
+
+        /* Skip the suite if we don't have an available implementation */
+        if (!potential_match->available) {
+            return false;
+        }
+
+        /* Make sure the cipher is valid for available certs */
+        if (s2n_is_cipher_suite_valid_for_auth(conn, potential_match) != S2N_SUCCESS) {
+            return false;
+        }
+
+        /* TLS 1.3 does not include key exchange in cipher suites */
+        if (potential_match->minimum_required_tls_version < S2N_TLS13) {
+            /* If the kex is not supported continue to the next candidate */
+            bool kex_supported = false;
+            POSIX_GUARD_RESULT(s2n_kex_supported(potential_match, conn, &kex_supported));
+            if (!kex_supported) {
+                return false;
+            }
+
+            /* If the kex is not configured correctly continue to the next candidate */
+            if (s2n_result_is_error(s2n_configure_kex(potential_match, conn))) {
+                return false;
+            }
+        }
+
+        /**
+        *= https://tools.ietf.org/rfc/rfc8446#section-4.2.11
+        *# The server MUST ensure that it selects a compatible PSK
+        *# (if any) and cipher suite.
+        **/
+        if (conn->psk_params.chosen_psk != NULL) {
+            if (potential_match->prf_alg != conn->psk_params.chosen_psk->hmac_alg) {
+                return false;
+            }
+        }
+
+        return true;
+}
+
+/**
+ * @brief Same as s2n_wire_ciphers_contain() except that it returns the match's index in the client wire.
+ * 
+ * @param match Cipher suite IANA which we wish to find a match for
+ * @param wire Client wire
+ * @param count Number of entries
+ * @param cipher_suite_len Length offset
+ * @return int Index of the match in the client wire. If a match was not found then -1 is returned.
+ */
+static int s2n_wire_ciphers_has_server_cipher_at(const uint8_t *match, const uint8_t *wire, uint32_t count, uint32_t cipher_suite_len)
```

Comment:
Can we somehow avoid duplicating s2n_wire_ciphers_contain? The logic there is already hard to follow.

Reply:
I'm going to leave this function as is. For the existing one, I can use this new function to determine if a match has been found. This reduces code duplication and removes an `memcmp` call